seed/analysis_pipelines/better.py

```diff
@@ -0,0 +1,448 @@
+                    analysis_property_view_id=analysis_property_view.id,
+                    user_message=f'Error preparing better input: {error}',
+                    debug_message='',
+                )
+            continue
+
+        analysis_input_file = AnalysisInputFile(
+            content_type=AnalysisInputFile.BUILDINGSYNC,
+            analysis=analysis
+        )
+        analysis_input_file.file.save(f'{analysis_property_view.id}.xml', ContentFile(better_doc))
+        analysis_input_file.clean()
+        analysis_input_file.save()
+        input_file_paths.append(analysis_input_file.file.path)
+
+    if len(input_file_paths) == 0:
+        pipeline = BETTERPipeline(analysis.id)
+        message = 'No files were able to be prepared for the analysis'
+        pipeline.fail(message, logger, progress_data_key=progress_data.key)
+        # stop the task chain
+        raise StopAnalysisTaskChain(message)
+
+
+@shared_task(bind=True)
+@analysis_pipeline_task(Analysis.CREATING)
+def _finish_preparation(self, analysis_id, progress_data_key):
+    """A Celery task which finishes the preparation for BETTER analysis
+
+    :param analysis_id: int
+    :param progress_data_key: str
+    """
+    analysis = Analysis.objects.get(id=analysis_id)
+    analysis.status = Analysis.READY
+    analysis.save()
+
+    progress_data = ProgressData.from_key(progress_data_key)
+    progress_data.finish_with_success()
+
+
+# PREMISES_ID_NAME is the name of the custom ID used within a BuildingSync document
+# to link it to SEED's AnalysisPropertyViews
+PREMISES_ID_NAME = 'seed_analysis_property_view_id'
+
+
+def _build_better_input(analysis_property_view, meters):
+    """Constructs a BuildingSync document to be used as input for a BETTER analysis.
+    The function returns a tuple, the first value being the XML document as a byte
+    string. The second value is a list of error messages.
+
+    :param analysis_property_view: AnalysisPropertyView
+    :param meter: Meter
+    :returns: tuple(bytes, list[str])
+    """
+    # TODO Refine ID assignment
+
+    errors = []
+    property_state = analysis_property_view.property_state
+
+    if property_state.property_name is None:
+        errors.append('Linked PropertyState is missing a name')
+    if property_state.city is None:
+        errors.append('Linked PropertyState is missing a city')
+    if property_state.state is None:
+        errors.append('Linked PropertyState is missing a State')
+    if property_state.gross_floor_area is None:
+        errors.append('Linked PropertyState is missing a gross floor area')
+    for meter in meters:
+        for meter_reading in meter.meter_readings.all():
+            if meter_reading.reading is None:
+                errors.append(f'{meter}: MeterReading starting at {meter_reading.start_time} has no reading value')
+    if errors:
+        return None, errors
+
+    # clean inputs
+    # BETTER will default if eGRIDRegion is not explicitly set
```

Comment:
Does this mean better will set a default eGRIDRegion value if not set? Or BETTER will fail?

Reply:
Yeah BETTER has a default region it assigns if none are provided so it will not fail.